File: implementations/javascript/bin/ff3-stresstest.js
#!/usr/bin/env node

import { randomBytes } from 'crypto';
import {
  digits,
  hexLower,
  hexUpper,
  base36Lower,
  base36Upper,
  base62
} from '../lib/index.js';

const DEFAULT_ITERATIONS = 1000;
const DEFAULT_MIN_LENGTH = 6;
const DEFAULT_MAX_LENGTH = 20;
const DEFAULT_ALPHABETS = ['digits', 'hex-lower', 'base36-lower', 'base62'];

const ALPHABETS = new Map([
  ['digits', '0123456789'],
  ['hex-lower', '0123456789abcdef'],
  ['hex-upper', '0123456789ABCDEF'],
  ['base36-lower', '0123456789abcdefghijklmnopqrstuvwxyz'],
  ['base36-upper', '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'],
  ['base62', '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz']
]);

function printUsage() {
  console.log('FF3 Stress Test Tool\n');
  console.log('Usage: ff3-stresstest [OPTIONS] [ITERATIONS]\n');
  console.log('Options:');
  console.log('  --alphabets LIST      Comma-separated list (default: digits,hex-lower,base36-lower,base62)');
  console.log('  --min-length N        Minimum plaintext length (default: 6)');
  console.log('  --max-length N        Maximum plaintext length (default: 20)');
  console.log('  --quick               Run 100 iterations (fast test)');
  console.log('  --seed N              Random seed for reproducibility');
}

function parseArgs(argv) {
  const opts = {
    iterations: DEFAULT_ITERATIONS,
    alphabets: [...DEFAULT_ALPHABETS],
    minLength: DEFAULT_MIN_LENGTH,
    maxLength: DEFAULT_MAX_LENGTH,
    quick: false,
    seed: undefined,
    help: false
  };

  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    switch (arg) {
      case '--help':
      case '-h':
        opts.help = true;
        return opts;
      case '--alphabets':
        if (i + 1 >= argv.length) throw new Error('--alphabets requires a value');
        opts.alphabets = argv[++i].split(',').map(s => s.trim()).filter(Boolean);
        if (opts.alphabets.length === 0) throw new Error('--alphabets list cannot be empty');
        break;
      case '--min-length':
        if (i + 1 >= argv.length) throw new Error('--min-length requires a value');
        opts.minLength = parseInt(argv[++i], 10);
        break;
      case '--max-length':
        if (i + 1 >= argv.length) throw new Error('--max-length requires a value');
        opts.maxLength = parseInt(argv[++i], 10);
        break;
      case '--quick':
        opts.quick = true;
        break;
      case '--seed':
        if (i + 1 >= argv.length) throw new Error('--seed requires a value');
        opts.seed = BigInt(parseInt(argv[++i], 10));
        break;
      default:
        if (arg.startsWith('--')) {
          throw new Error(`Unknown option: ${arg}`);
        }
        const iters = parseInt(arg, 10);
        if (!Number.isInteger(iters) || iters <= 0) {
          throw new Error(`Invalid iterations value: ${arg}`);
        }
        opts.iterations = iters;
    }
  }

  if (opts.quick) {
    opts.iterations = 100;
  }

  return opts;
}

function validateOptions(opts) {
  if (opts.iterations <= 0) throw new Error('iterations must be greater than 0');
  if (opts.minLength < 2) throw new Error('--min-length must be at least 2');
  if (opts.maxLength < opts.minLength) throw new Error('--max-length must be greater than or equal to --min-length');
  for (const name of opts.alphabets) {
    if (!ALPHABETS.has(name)) {
      throw new Error(`Unknown alphabet: ${name}`);
    }
  }
}

function createCipherFactory(name) {
  switch (name) {
    case 'digits': return digits;
    case 'hex-lower': return hexLower;
    case 'hex-upper': return hexUpper;
    case 'base36-lower': return base36Lower;
    case 'base36-upper': return base36Upper;
    case 'base62': return base62;
    default: throw new Error(`Unknown alphabet: ${name}`);
  }
}

function toHex(buffer) {
  return Buffer.from(buffer).toString('hex');
}

function createRng(seed) {
  if (seed === undefined) {
    return {
      bytes: n => randomBytes(n),
      int: (min, max) => Math.floor(Math.random() * (max - min + 1)) + min
    };
  }

  let state = seed & ((1n << 48n) - 1n);
  const modulus = 1n << 48n;
  const multiplier = 25214903917n;
  const increment = 11n;

  function next() {
    state = (state * multiplier + increment) % modulus;
    return Number(state >> 16n) & 0x7fffff;
  }

  return {
    bytes: n => {
      const buf = Buffer.allocUnsafe(n);
      for (let i = 0; i < n; i++) {
        buf[i] = next() & 0xff;
      }
      return buf;
    },
    int: (min, max) => {
      const range = max - min + 1;
      return min + (next() % range);
    }
  };
}

function stressAlphabet(name, charset, factory, opts, rng) {
  console.log(`Testing ${name}...`);
  console.log(`  Alphabet: ${charset} (radix ${charset.length})`);

  let passed = 0;
  let failed = 0;
  const interval = Math.max(1, Math.floor(opts.iterations / 10));

  for (let i = 0; i < opts.iterations; i++) {
    const key = rng.bytes(16);
    const tweak = rng.bytes(8);
    const length = rng.int(opts.minLength, opts.maxLength);
    const plaintext = generatePlaintext(charset, length, rng);

    try {
      const cipher = factory(key, tweak);
      const ciphertext = cipher.encrypt(plaintext);
      const decrypted = cipher.decrypt(ciphertext);
      if (decrypted !== plaintext) {
        failed += 1;
        printFailure(key, tweak, plaintext, ciphertext, decrypted, 'round-trip mismatch');
      } else {
        passed += 1;
      }
    } catch (err) {
      failed += 1;
      printFailure(key, tweak, plaintext, '', null, err.message);
    }

    if ((i + 1) % interval === 0 || i + 1 === opts.iterations) {
      const percent = Math.floor(((i + 1) * 100) / opts.iterations);
      console.log(`  Progress: ${i + 1}/${opts.iterations} (${percent}%)`);
    }
  }

  console.log(`  Passed: ${passed}/${opts.iterations}`);
  console.log(`  Failed: ${failed}/${opts.iterations}\n`);
  return { tests: opts.iterations, failures: failed };
}

function generatePlaintext(charset, length, rng) {
  const chars = [];
  for (let i = 0; i < length; i++) {
    const idx = rng.int(0, charset.length - 1);
    chars.push(charset[idx]);
  }
  return chars.join('');
}

function printFailure(key, tweak, plain, cipher, decrypted, detail) {
  console.log('  Round-trip failed:');
  console.log(`    Key: ${toHex(key)}`);
  console.log(`    Tweak: ${toHex(tweak)}`);
  console.log(`    Plaintext: "${plain}"`);
  if (cipher) {
    console.log(`    Ciphertext: "${cipher}"`);
  }
  if (decrypted) {
    console.log(`    Decrypted: "${decrypted}"`);
  }
  if (detail) {
    console.log(`    Detail: ${detail}`);
  }
}

function main() {
  let opts;
  try {
    opts = parseArgs(process.argv.slice(2));
  } catch (err) {
    console.error(`Error: ${err.message}`);
    process.exit(1);
  }

  if (opts.help) {
    printUsage();
    return;
  }

  try {
    validateOptions(opts);
  } catch (err) {
    console.error(`Error: ${err.message}`);
    process.exit(1);
  }

  const rng = createRng(opts.seed);

  console.log('FF3 Stress Test v1.0');
  console.log('====================\n');
  console.log('Warning: FF3 was withdrawn by NIST; run for education and research only.\n');
  console.log('Test configuration');
  console.log(`  Iterations per alphabet: ${opts.iterations}`);
  console.log('  Random key/tweak generation: enabled');
  console.log(`  String length range: ${opts.minLength}-${opts.maxLength} characters`);
  console.log(`  Alphabets: ${opts.alphabets.join(', ')}\n`);

  const start = Date.now();
  let totalTests = 0;
  let totalFailures = 0;

  for (const name of opts.alphabets) {
    const charset = ALPHABETS.get(name);
    const factory = createCipherFactory(name);
    const result = stressAlphabet(name, charset, factory, opts, rng);
    totalTests += result.tests;
    totalFailures += result.failures;
  }

  const duration = Date.now() - start;
  console.log('Summary');
  console.log(`  Total tests: ${totalTests}`);
  console.log(`  Failures: ${totalFailures}`);
  console.log(`  Duration: ${duration} ms`);
  if (duration > 0) {
    const throughput = (totalTests * 1000) / duration;
    console.log(`  Throughput: ${throughput.toFixed(2)} tests/sec`);
  }
  if (totalFailures === 0) {
    console.log('  Result: all stress tests passed');
  } else {
    console.log('  Result: failures detected');
    process.exit(1);
  }
}

main();

File: implementations/rust/src/bin/ff3-stresstest.rs
use fpe_ff3::{base36_lower, base36_upper, base62, digits, hex_lower, hex_upper, FF3};
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};
use std::env;
use std::time::Instant;

const DEFAULT_ITERATIONS: usize = 1000;
const DEFAULT_MIN_LEN: usize = 6;
const DEFAULT_MAX_LEN: usize = 20;
const DEFAULT_ALPHABETS: [&str; 4] = ["digits", "hex-lower", "base36-lower", "base62"];

#[derive(Clone)]
struct Options {
    iterations: usize,
    alphabets: Vec<String>,
    min_len: usize,
    max_len: usize,
    quick: bool,
    seed: Option<u64>,
    help: bool,
}

impl Default for Options {
    fn default() -> Self {
        Self {
            iterations: DEFAULT_ITERATIONS,
            alphabets: DEFAULT_ALPHABETS.iter().map(|s| s.to_string()).collect(),
            min_len: DEFAULT_MIN_LEN,
            max_len: DEFAULT_MAX_LEN,
            quick: false,
            seed: None,
            help: false,
        }
    }
}

struct AlphabetInfo {
    name: &'static str,
    charset: &'static str,
    factory: fn(&[u8], &[u8]) -> Result<FF3, String>,
}

fn main() {
    if let Err(err) = run() {
        eprintln!("Error: {}", err);
        std::process::exit(1);
    }
}

fn run() -> Result<(), String> {
    let opts = parse_args(env::args().skip(1))?;

    if opts.help {
        print_usage();
        return Ok(());
    }

    let mut opts = opts;
    if opts.quick {
        opts.iterations = 100;
    }

    validate_options(&opts)?;

    let mut rng = match opts.seed {
        Some(seed) => StdRng::seed_from_u64(seed),
        None => StdRng::from_entropy(),
    };

    println!("FF3 Stress Test v1.0");
    println!("====================\n");
    println!("Warning: FF3 was withdrawn by NIST; run for education and research only.\n");
    println!("Test configuration");
    println!("  Iterations per alphabet: {}", opts.iterations);
    println!("  Random key/tweak generation: enabled");
    println!("  String length range: {}-{} characters", opts.min_len, opts.max_len);
    println!("  Alphabets: {}\n", opts.alphabets.join(", "));

    let alph_infos = opts
        .alphabets
        .iter()
        .map(|name| alphabet_info(name))
        .collect::<Result<Vec<_>, _>>()?;

    let start = Instant::now();
    let mut total_tests = 0usize;
    let mut total_failures = 0usize;

    for info in alph_infos {
        let (tests, failures) = stress_alphabet(&info, &opts, &mut rng);
        total_tests += tests;
        total_failures += failures;
    }

    let duration = start.elapsed();
    println!("Summary");
    println!("  Total tests: {}", total_tests);
    println!("  Failures: {}", total_failures);
    println!("  Duration: {} ms", duration.as_millis());
    if duration.as_millis() > 0 {
        let throughput = (total_tests as f64 * 1000.0) / duration.as_millis() as f64;
        println!("  Throughput: {:.2} tests/sec", throughput);
    }
    if total_failures == 0 {
        println!("  Result: all stress tests passed");
        Ok(())
    } else {
        println!("  Result: failures detected");
        Err(String::from("failures detected"))
    }
}

fn stress_alphabet(info: &AlphabetInfo, opts: &Options, rng: &mut StdRng) -> (usize, usize) {
    println!("Testing {}...", info.name);
    println!("  Alphabet: {} (radix {})", info.charset, info.charset.len());

    let mut passed = 0usize;
    let mut failed = 0usize;
    let interval = usize::max(1, opts.iterations / 10);

    for iter in 0..opts.iterations {
        let key = random_bytes(rng, 16);
        let tweak = random_bytes(rng, 8);
        let length = rng.gen_range(opts.min_len..=opts.max_len);
        let plaintext = generate_plaintext(rng, info.charset, length);

        match (info.factory)(&key, &tweak) {
            Ok(cipher) => match cipher.encrypt(&plaintext, None) {
                Ok(ciphertext) => match cipher.decrypt(&ciphertext, None) {
                    Ok(decrypted) => {
                        if decrypted == plaintext {
                            passed += 1;
                        } else {
                            failed += 1;
                            print_failure(&key, &tweak, &plaintext, Some(&ciphertext), Some(&decrypted), Some("round-trip mismatch"));
                        }
                    }
                    Err(err) => {
                        failed += 1;
                        print_failure(&key, &tweak, &plaintext, Some(&ciphertext), None, Some(&format!("decryption error: {err}")));
                    }
                },
                Err(err) => {
                    failed += 1;
                    print_failure(&key, &tweak, &plaintext, None, None, Some(&format!("encryption error: {err}")));
                }
            },
            Err(err) => {
                failed += 1;
                print_failure(&key, &tweak, &plaintext, None, None, Some(&format!("cipher init error: {err}")));
            }
        }

        if (iter + 1) % interval == 0 || iter + 1 == opts.iterations {
            let percent = ((iter + 1) * 100) / opts.iterations;
            println!("  Progress: {}/{} ({}%)", iter + 1, opts.iterations, percent);
        }
    }

    println!("  Passed: {}/{}", passed, opts.iterations);
    println!("  Failed: {}/{}\n", failed, opts.iterations);
    (opts.iterations, failed)
}

fn random_bytes(rng: &mut StdRng, length: usize) -> Vec<u8> {
    (0..length).map(|_| rng.gen()).collect()
}

fn generate_plaintext(rng: &mut StdRng, alphabet: &str, length: usize) -> String {
    let chars: Vec<char> = alphabet.chars().collect();
    (0..length)
        .map(|_| chars[rng.gen_range(0..chars.len())])
        .collect()
}

fn print_failure(key: &[u8], tweak: &[u8], plain: &str, cipher: Option<&str>, decrypted: Option<&str>, detail: Option<&str>) {
    println!("  Round-trip failed:");
    println!("    Key: {}", hex(key));
    println!("    Tweak: {}", hex(tweak));
    println!("    Plaintext: \"{}\"", plain);
    if let Some(ct) = cipher {
        println!("    Ciphertext: \"{}\"", ct);
    }
    if let Some(pt) = decrypted {
        println!("    Decrypted: \"{}\"", pt);
    }
    if let Some(msg) = detail {
        println!("    Detail: {}", msg);
    }
}

fn hex(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{:02x}", b)).collect()
}

fn parse_args<I>(args: I) -> Result<Options, String>
where
    I: IntoIterator<Item = String>,
{
    let mut opts = Options::default();
    let mut iter = args.into_iter();
    while let Some(arg) = iter.next() {
        match arg.as_str() {
            "--help" | "-h" => {
                opts.help = true;
                return Ok(opts);
            }
            "--alphabets" => {
                let value = iter.next().ok_or("--alphabets requires a value")?;
                let list: Vec<String> = value
                    .split(',')
                    .map(|s| s.trim().to_lowercase())
                    .filter(|s| !s.is_empty())
                    .collect();
                if list.is_empty() {
                    return Err(String::from("--alphabets list cannot be empty"));
                }
                opts.alphabets = list;
            }
            "--min-length" => {
                let value = iter.next().ok_or("--min-length requires a value")?;
                opts.min_len = value.parse().map_err(|_| "invalid value for --min-length")?;
            }
            "--max-length" => {
                let value = iter.next().ok_or("--max-length requires a value")?;
                opts.max_len = value.parse().map_err(|_| "invalid value for --max-length")?;
            }
            "--quick" => opts.quick = true,
            "--seed" => {
                let value = iter.next().ok_or("--seed requires a value")?;
                let seed: u64 = value.parse().map_err(|_| "invalid value for --seed")?;
                opts.seed = Some(seed);
            }
            other if other.starts_with('-') => return Err(format!("Unknown option: {}", other)),
            value => {
                let iterations: usize = value.parse().map_err(|_| "invalid iterations value")?;
                if iterations == 0 {
                    return Err(String::from("iterations must be greater than 0"));
                }
                opts.iterations = iterations;
            }
        }
    }
    Ok(opts)
}

fn validate_options(opts: &Options) -> Result<(), String> {
    if opts.iterations == 0 {
        return Err(String::from("iterations must be greater than 0"));
    }
    if opts.min_len < 2 {
        return Err(String::from("--min-length must be at least 2"));
    }
    if opts.max_len < opts.min_len {
        return Err(String::from("--max-length must be greater than or equal to --min-length"));
    }
    for name in &opts.alphabets {
        alphabet_info(name)?;
    }
    Ok(())
}

fn alphabet_info(name: &str) -> Result<AlphabetInfo, String> {
    match name {
        "digits" => Ok(AlphabetInfo { name: "digits", charset: "0123456789", factory: digits }),
        "hex-lower" => Ok(AlphabetInfo { name: "hex-lower", charset: "0123456789abcdef", factory: hex_lower }),
        "hex-upper" => Ok(AlphabetInfo { name: "hex-upper", charset: "0123456789ABCDEF", factory: hex_upper }),
        "base36-lower" => Ok(AlphabetInfo { name: "base36-lower", charset: "0123456789abcdefghijklmnopqrstuvwxyz", factory: base36_lower }),
        "base36-upper" => Ok(AlphabetInfo { name: "base36-upper", charset: "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ", factory: base36_upper }),
        "base62" => Ok(AlphabetInfo { name: "base62", charset: "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz", factory: base62 }),
        other => Err(format!("Unknown alphabet: {}", other)),
    }
}

fn print_usage() {
    println!("FF3 Stress Test Tool\n");
    println!("Usage: ff3-stresstest [OPTIONS] [ITERATIONS]\n");
    println!("Options:");
    println!("  --alphabets LIST      Comma-separated list (default: digits,hex-lower,base36-lower,base62)");
    println!("  --min-length N        Minimum plaintext length (default: 6)");
    println!("  --max-length N        Maximum plaintext length (default: 20)");
    println!("  --quick               Run 100 iterations (fast test)");
    println!("  --seed N              Random seed for reproducibility");
}
